# Combining DataFrames With Pandas - Lab

## Introduction

In this lab, you'll gain practice combining DataFrames through concatenation.  You'll also practice executing various types of joins to selectively combine the information stored in the tables!

## Objectives

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)


## Getting Started

You'll start with a quick section to gain practice with concatenating datasets using `pd.concat()`.

## Concatenating DataFrames

Run the cell below to create some sample DataFrames for us to work with.  

In [1]:
import pandas as pd
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                    index=[0, 1, 2, 3])


df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                    index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'], 
                    'D': ['D8', 'D9', 'D10', 'D11']},
                    index=[8, 9, 10, 11])

Now that you have multiple DataFrames to work with, you can execute a concatenation to join them together.  

In the cell below, concatenate the 3 DataFrames together using the appropriate function.  

In [2]:
combined_df = pd.concat([df1, df2, df3])
combined_df

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

**_EXPECTED OUTPUT:_**

<img src="images/er1.png">

## Setting Join Conditions With Concatenation

You can also specify if the concatenation is an **_Outer Join_** or an **_Inner Join_**.  Next, you'll execute an inner join. Before you do, you need to create another table that contains some overlapping index values with a DataFrame that already exists.  

Run the cell below to create the new DataFrame.

In [3]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'],
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                    index=[2, 3, 6, 7])

Now, in the cell below, use the `pd.concat()` method to join DataFrames 1 and 4.  However, this time, specify that the `join` is `'inner'`, and `axis=1`. 

In [4]:
df1_and_4 = pd.concat([df1,df4], axis = 1, join='inner')
df1_and_4

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

**_EXPECTED OUTPUT:_**

<img src='images/er2.png'>

You'll notice that in this case, the results returned contain only the rows with indexes that exist in both tables--rows 2 and 3.  The resulting table contains the values for each column in both tables for the rows.  

Note that there are many, many ways that you can make full use of the `concat()` functionality in pandas to join DataFrames together--these are just a few of the most common examples pulled from the pandas documentation on the subject.  For a full view of all the ways you can use `pd.concat()`, see the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html)!

## Loading In Data
Now, it's time to move on to working with the Hearthstone cards database.  This database contains information on cards from the popular game, [Hearthstone](https://playhearthstone.com/en-us/)! For full information on the dataset, see the  [kaggle page](https://www.kaggle.com/jeradrose/hearthstone-cards) for this dataset. 

This database consists of the following tables:

* _cards_
* *dust_costs*
* _entourages_
* _mechanics_
* *play_requirements*

Many of rows in each table--but not all--correspond to the same cards. As such, each table contains a column called `card_id` which acts as a **_Primary Key_** for each table.  You'll make use of these keys to **_join_** the different tables together into a single DataFrame.  You'll also experiment with different types of joins to help us decide exactly what information you wish to combine.  

Simply run the cell below to import the tables from the database as DataFrames.

In [35]:
cards_df = pd.read_csv('cards.csv')
dust_df = pd.read_csv('dust.csv')
entourages_df = pd.read_csv('entourages.csv')
mechanics_df = pd.read_csv('mechanics.csv')
play_requirements_df = pd.read_csv('play_requirements.csv')

Great.  Now, let's set the correct column, `card_id`, as the index column for each of these tables, and then display each to ensure that everything is as expected.  

For each of the dataframes you created in the cell above, call the `.set_index()` method and pass in `card_id`.  Also set `inplace=True`.  Then, display the head of each respective DataFrame to ensure everything worked.  

**_NOTE:_** Since you are performing this operation in place, running any cell a second time will result in pandas throwing an error.  If you need to run something a second time, restart the kernel using the jupyter notebook menu at the top of the page.  

In [36]:
#cards_df.set_index('card_id', inplace = True)
cards_df

card_id player_class         type  \
0                        KARA_00_07         MAGE        SPELL   
1                         NEW1_008a        DRUID        SPELL   
2                         BRM_010t2        DRUID       MINION   
3                            AT_132      NEUTRAL       MINION   
4                            OG_141      NEUTRAL       MINION   
5                           CRED_17      NEUTRAL       MINION   
6                           OG_337e      NEUTRAL  ENCHANTMENT   
7                         BRMA08_1H      NEUTRAL         HERO   
8                          GVG_011a      NEUTRAL  ENCHANTMENT   
9                       NAX8_02H_TB      NEUTRAL   HERO_POWER   
10                           AT_131      NEUTRAL       MINION   
11                         CS2_122e      NEUTRAL  ENCHANTMENT   
12                      LOEA05_02ha      NEUTRAL   HERO_POWER   
13                CS2_102_H1_AT_132      WARRIOR   HERO_POWER   
14                         CFM_334e      NEUTRAL  ENCHANTMENT   
15                          EX1_583      NEUTRAL       MINION   
16           TB_PickYourFate_11rand      NEUTRAL        SPELL   
17                         LOE_007t      WARLOCK        SPELL   
18                          CFM_095      NEUTRAL       MINION   
19                          BRM_018      PALADIN       MINION   
20                         PRO_001a      NEUTRAL        SPELL   
21                        BRMA10_5H      NEUTRAL       MINION   
22                         BRMA11_1      NEUTRAL         HERO   
23                           OG_027       SHAMAN        SPELL   
24                          NAX1_01      NEUTRAL         HERO   
25                          GVG_030        DRUID       MINION   
26                          CFM_094      WARLOCK        SPELL   
27                      DS1h_292_H1       HUNTER   HERO_POWER   
28                          CS2_227      NEUTRAL       MINION   
29                          KAR_062      NEUTRAL       MINION   
...                             ...          ...          ...   
2789                       LOEA12_2      NEUTRAL   HERO_POWER   
2790                         AT_084      NEUTRAL       MINION   
2791                     KARA_00_03         MAGE         HERO   
2792                       BRMA16_3      NEUTRAL        SPELL   
2793                        tt_010a         MAGE       MINION   
2794                        EX1_559         MAGE       MINION   
2795                        EX1_624       PRIEST        SPELL   
2796                        OG_044a        DRUID       MINION   
2797                         AT_044        DRUID        SPELL   
2798                  AT_132_SHAMAN       SHAMAN   HERO_POWER   
2799                     CFM_621t29      NEUTRAL        SPELL   
2800                       EX1_355e      PALADIN  ENCHANTMENT   
2801                      LOEA09_3d      NEUTRAL   HERO_POWER   
2802                  TB_Coopv3_105      NEUTRAL       MINION   
2803                       BRMA13_4      NEUTRAL   HERO_POWER   
2804                       EX1_345t       PRIEST       MINION   
2805                        EX1_009      NEUTRAL       MINION   
2806                        CRED_32      NEUTRAL       MINION   
2807                       FP1_030e      NEUTRAL  ENCHANTMENT   
2808        TB_PickYourFate_10_Ench      NEUTRAL  ENCHANTMENT   
2809             KARA_05_01hpheroic      NEUTRAL   HERO_POWER   
2810                       BRMA10_3      NEUTRAL   HERO_POWER   
2811  TB_PickYourFate_10_EnchMinion      NEUTRAL  ENCHANTMENT   
2812                       NAX3_02H      NEUTRAL   HERO_POWER   
2813                        AT_083e      NEUTRAL  ENCHANTMENT   
2814                        AT_066e      WARRIOR  ENCHANTMENT   
2815                        CS2_031         MAGE        SPELL   
2816                        OG_195e      NEUTRAL  ENCHANTMENT   
2817                       XXX_111e      NEUTRAL  ENCHANTMENT   
2818                        NAX3_01      NEUTRAL         HERO  

In [37]:
dust_df.set_index('card_id', inplace = True)

In [38]:
entourages_df.set_index('card_id', inplace = True)

In [39]:
#mechanics_df = mechanics_df.set_index('card_id', inplace = True)
mechanics_df.set_index('card_id', inplace = True)

In [40]:
#play_requirements_df = play_requirements_df.set_index('card_id', inplace = True)
play_requirements_df.set_index('card_id', inplace = True)

## Executing Joins

Now that you have the tables loaded correctly, we're going to execute some joins. There are 4 different kinds of joins, which can best be visualized with venn diagrams:

<img src='images/joins.png'>

In these diagrams, each circle represents a DataFrame or SQL Table.  The left table is the table you are working with, and the right table is the table you want to join to the table you are working with.  You'll start by executing the most common type of join, an **_Inner Join_**.

In the cell below, join `cards_df` with `mechanics_df` using the built-in `.join()` method on the `cards_df` object. 

Pass in the following parameters:
* the table you want to join with, `mechanics_df`
* The `how` parameter set to the type of join you want, `'inner'`

In [34]:
cards_df = pd.join([cards_df, mechanics_df], axis=1, join='inner')
cards_df

mechanic
card_id                                
AT_132                        BATTLECRY
GVG_011a            TAG_ONE_TURN_EFFECT
EX1_583                       BATTLECRY
LOE_007t                      EVIL_GLOW
LOE_007t             ImmuneToSpellpower
CFM_095                     DEATHRATTLE
BRM_018                       BATTLECRY
GVG_030                      CHOOSE_ONE
GVG_030                           TAUNT
CS2_227                            AURA
KAR_062                       BATTLECRY
CFM_712_t17                  JADE_GOLEM
BRM_011t            TAG_ONE_TURN_EFFECT
BRMA10_3H                  AI_MUST_PLAY
KAR_710m                          TAUNT
EX1_295o            TAG_ONE_TURN_EFFECT
OG_158                      DEATHRATTLE
LOE_110t                        TOPDECK
LOE_110t             ImmuneToSpellpower
BRMA13_4H                  AI_MUST_PLAY
EX1_145o            TAG_ONE_TURN_EFFECT
LOEA09_2e           TAG_ONE_TURN_EFFECT
EX1_570e            TAG_ONE_TURN_EFFECT
GVG_069                       BATTLECRY
hexfrog                           TAUNT
OG_145                    DIVINE_SHIELD
OG_145                            TAUNT
AT_090                          INSPIRE
AT_060                           SECRET
TB_MechWar_Minion1                TAUNT
...                                 ...
EX1_284                       BATTLECRY
LOE_038                            AURA
AT_083                          INSPIRE
TB_CoOpv3_BOSS2e                ENRAGED
GVG_009                       BATTLECRY
CFM_338                       BATTLECRY
GVG_097                       BATTLECRY
GVG_097                           TAUNT
CFM_852                       BATTLECRY
LOE_020                       BATTLECRY
CFM_781                         STEALTH
GVG_090                       BATTLECRY
GVG_120                       BATTLECRY
EX1_055o            TAG_ONE_TURN_EFFECT
CFM_321                       BATTLECRY
EX1_082                       BATTLECRY
FP1_017                            AURA
FP1_001                     DEATHRATTLE
AT_089                          INSPIRE
CFM_646                     DEATHRATTLE
AT_084                        BATTLECRY
OG_044a                          CHARGE
OG_044a                           TAUNT
LOEA09_3d                  AI_MUST_PLAY
BRMA13_4                   AI_MUST_PLAY
EX1_009                         ENRAGED
BRMA10_3                   AI_MUST_PLAY
NAX3_02H                   AI_MUST_PLAY
AT_083e             TAG_ONE_TURN_EFFECT
CS2_031                          FREEZE

[1079 rows x 1 columns]

Examine the output from the cell above and compare it to the original output of both the `cards_df` and `mechanics_df` DataFrame heads you displayed earlier.  Notice how it this now combines the columns from both?

**_Question_**

If you inspect the original `cards_df` DataFrame, you'll notice that it contains  2,819 records.  The result of our inner join, `cards_with_mechanics_df`, conntains only 1079 rows.  Why?

Write your answer below this line:
________________________________________________________________________________

First performed an inner join, which only includes records that are present in both tables.  Although there were 2819 records in the left table, there were only 1079 records that existed in both tables, which are what you see in the resulting dataframe. 

## Other Types of Joins

By default, the `.join()` method performs a left join if no parameter is passed in for `how=`.  In the cell below, perform a **_Left Join_** of `cards_with_mechanics_df` and `play_requirements_df`, with `cards_with_mechanics_df` as the left table.  

Then, display `left_join_df` to inspect our results. 

In [ ]:
left_join_df = None
left_join_df

Note that the results of this sort of join are dependent upon the position of each table--if you were to make `cards_with_mechanics_df` the right table and `play_requirements_df` the left table and then perform a **_Right Join_**, our results would be the same. 

**_Question:_**

Describe what was included from each table in this join.

Write your answer below this line:
________________________________________________________________________________

Every record from `cards_with_mechanics_df`, as well as any records from `play_requirements_df` that have matching index values with a record from the left table.  

#### Outer Joins

In the cell below, perform an outer join between `cards_df` and `dust_df`. Since these tables contain columns with the same name, we'll need to specify a suffix for at least one of them, so that the column can be renamed to avoid a naming collision. 

During this join, set the `rsuffix` parameter to `_dust`

In [ ]:
outer_join_df = None
outer_join_df

Inspect the output above.  Note that the naming collision has been avoided by renaming the `cost` column from the right table to `cost_dust`.  

## Summary

In this lab, you learned how to:

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)